### PR TITLE
Add some missing grouped_df methods

### PR DESCRIPTION
**Pull request by hadley**

### CODE REVIEW

File: tests/testthat/test-grouped-df.r [hunk 1]
```diff
@@ -6,3 +6,45 @@ test_that("can selectively ungroup", {
...
 })
+
+
+test_that("[ method can remove grouping vars", {
+  df <- tibble(x = 1, y = 2, z = 3)
+  gf <- group_by(df, x, y)
+
+  expect_equal(gf, gf)
+  expect_equal(gf[1], group_by(df[1], x))
+  expect_equal(gf[3], df[3])
```
Comment (lionel-) [inline tests/testthat/test-grouped-df.r]: Do we want to remove the grouped-df class here? I think we should still pursue the idea of frame prototypes, to make things easier in the future. In that case, `gf[3]` would return a zero-grouping-var grouped-df.

File: R/grouped-df.r [hunk 2]
```diff
@@ -204,20 +204,38 @@ as_tibble.grouped_df <- function(x, ...) {
...
   } else {
-    grouped_df(y, group_names, group_by_drop_default(x))
+    out
   }
 }
 
+#' @export
+`[<-.grouped_df` <- function(x, i, j, ..., value) {
+  out <- NextMethod()
+  grouped_df(out, intersect(names(out), group_vars(x)), group_by_drop_default(x))
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/grouped-df.r]: After a bit of experimentation, I think that's too complicated — instead, specific dplyr methods will avoid `[.grouped_df` in favour of their own custom restoration, as in #4715

File: R/grouped-df.r [hunk 3]
```diff
@@ -204,20 +204,38 @@ as_tibble.grouped_df <- function(x, ...) {
 #' @importFrom tibble is_tibble
 #' @export
 `[.grouped_df` <- function(x, i, j, drop = FALSE) {
-  y <- NextMethod()
+  out <- NextMethod()
 
-  if (isTRUE(drop) && !is_tibble(y)) {
-    return(y)
+  if (isTRUE(drop) && !is.data.frame(out)) {
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/grouped-df.r]: The intent of this code is protect against the matrix indexing case, e.g. `df[cbind(1, 2)]` which returns a vector, not a data frame. I'll reconsider the logic in light of your comments.

File: R/grouped-df.r [hunk 4]
```diff
@@ -204,20 +204,38 @@ as_tibble.grouped_df <- function(x, ...) {
...
+    return(out)
   }
 
-  group_names <- group_vars(x)
-  if (!all(group_names %in% names(y))) {
-    tbl_df(y)
+  groups <- intersect(names(out), group_vars(x))
+  grouped_df(out, groups, group_by_drop_default(x))
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/grouped-df.r]: I think that's a reasonable optimisation; I'll do that.

File: R/grouped-df.r [hunk 5]
```diff
@@ -204,20 +204,38 @@ as_tibble.grouped_df <- function(x, ...) {
...
+#' @export
+`[<-.grouped_df` <- function(x, i, j, ..., value) {
+  out <- NextMethod()
+  grouped_df(out, intersect(names(out), group_vars(x)), group_by_drop_default(x))
+}
+
+#' @export
+`[[<-.grouped_df` <- function(x, ..., value) {
+  out <- NextMethod()
+  grouped_df(out, intersect(names(out), group_vars(x)), group_by_drop_default(x))
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/grouped-df.r]: The key challenge here is turning `...` into an index that I can apply to `group_vars()`. There's so much flexibility baked into `[[` it's really hard to cover every case.

File: tests/testthat/test-grouped-df.r (hunk 1 above)
Comment (hadley) [reply tests/testthat/test-grouped-df.r]: We don't currently support zero-col grouping, so I think that would be a big change, and out of scope for this PR.

File: R/grouped-df.r (hunk 4 above)
[the remark this answers is not in the document]
Comment (hadley) [reply R/grouped-df.r]: Oh but that only works if `i` is missing as if `df[, 1:2]`, or `nargs() == 2` as in `df[1:2]`

**Pull request merged**